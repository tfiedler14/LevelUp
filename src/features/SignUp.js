import * as React from 'react';
import { connect } from 'react-redux';
import { View, ImageBackground } from 'react-native';
import { setLocation } from '../logic/location/actions';
import { setAuth } from '../logic/auth/actions';
import { firebaseApp } from '../../Const';
import { Field, reduxForm } from 'redux-form';
import { compose } from 'redux';
import { WrappedTextInput } from '../shared-components/FormField';
import EStyleSheet from 'react-native-extended-stylesheet';
import { Button, Card } from 'react-native-paper';
import { FormHeader } from '../shared-components/FormHeader';
import { putData } from '../logic/data/actions';

export const SignUp = ({ setLocation, handleSubmit, setAuthentication, putData }) => {
  return (
    <ImageBackground source={require('../../assets/images/darkverylowopacityshapes.png')} style={{height: '100%', width: '100%'}}>
    <View style={{paddingTop: 100}}>
    <Card style={styles.card}>
      <View>
        <FormHeader title={'Sign Up'} />
        <Field
          name="email"
          id="email"
          props={{
            title: 'Email', textContentType: 'emailAddress', autoCompleteType: 'email'
          }}
          component={WrappedTextInput}
        />
        <Field
          name="password"
          id="password"
          props={{
            title: 'Password',
            textContentType: 'password',
            autoCompleteType: 'password',
            password: true
          }}
          component={WrappedTextInput}
        />
        <Field
          name="cpassword"
          id="cpassword"
          props={{
            title: 'Confirm Password',
            textContentType: 'password',
            autoCompleteType: 'password',
            password: true
          }}
          component={WrappedTextInput}
        />
        <View style={styles.buttons}>
          <Button
            color="#de3c57"
            uppercase={false}
            mode="contained"
            onPress={handleSubmit(values => {
              handleSignUp(values, setLocation, setAuthentication, putData);
            })}>
            Sign Up
          </Button>
          <Button
            color="white"
            uppercase={false}
            mode="text"
            onPress={() => setLocation('signin')}>
            Sign In
          </Button>
        </View>
      </View>
    </Card>
    </View>
    </ImageBackground>
  );
};

export const validate = values => {
  const errors = {};

  if (!values.email) {
    errors.email = 'Required';
  } else if (!/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(values.email)) {
    errors.email = 'Invalid email address';
  }

  if (!values.password) {
    errors.password = 'Required';
  } else if (values.password.length < 6 || values.password.length > 16) {
    errors.password = 'Must be 6-16 characters';
  }

  if (values.cpassword !== values.password) {
    errors.cpassword = 'Passwords must match';
  }

  return errors;
};

const handleSignUp = (values, setLocation, setAuthentication, putData) => {
  firebaseApp
    .auth()
    .createUserWithEmailAndPassword(values.email, values.password)
    .then(response => {
      setAuthentication({ loggedIn: true, email: values.email, uid: response.user.uid });
      putData('https://levelup-10cfc.firebaseio.com/users/' + response.user.uid + '.json', {
        quests: ['empty'],
        profile: { brows: "type-1", eyes: "type-1", hairColor: "type-1", hairStyle: "type-1", headShape: "type-1", mouth: "type-1" ,name: "none", nose: "type-1", skinColor: "tan" },
        skills: ['empty'],
        attributes: {fitness: {exp: 0, level: 1}, academics: {exp: 0, level: 1}, crafts: {exp: 0, level: 1}, community: {exp: 0, level: 1}, mental: {exp: 0, level: 1}, hobby: {exp: 0, level: 1}}
      }, 'profile', 'profile');
    })
    .catch(error => {
      console.log('Failed to create user.');
    });
};

const styles = EStyleSheet.create({
  card: {
    padding: '1rem',
    margin: '1rem',
    backgroundColor: '#555'
    //  border: 'none',
  },

  buttons: {
    marginTop: '1rem',
    marginLeft: '2rem',
    marginRight: '2rem'
  }
});

export const mapStateToProps = state => {
  return {
    location: state.location
  };
};

export const mapDispatchToProps = dispatch => {
  return {
    setLocation: location => {
      dispatch(setLocation(location));
    },
    setAuthentication: auth => {
      dispatch(setAuth(auth));
    },
    putData: (target, data, redirect, type) => {
      dispatch(putData(target, data, redirect, type));
    }
  };
};

export default compose(
  connect(
    mapStateToProps,
    mapDispatchToProps
  ),
  reduxForm({ form: 'sign-up-form', validate })
)(SignUp);
