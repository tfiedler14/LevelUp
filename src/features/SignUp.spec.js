import React from 'react';
import { shallow } from 'enzyme';
import { mapDispatchToProps, mapStateToProps, SignUp, validate } from './SignUp';

describe('SignIn', () => {
  it('should render sign up correctly', () => {
    const component = shallow(
      <SignUp
        setLocation={jest.fn()}
        handleSubmit={jest.fn()}
        setAuthentication={jest.fn()}
        putData={jest.fn()}
      />
    );

    expect(component).toMatchSnapshot();
  });
});

describe('mapDispatchToProps', () => {
  it('should dispatch putData when putData is called', () => {
    const dispatch = jest.fn();
    const props = mapDispatchToProps(dispatch);
    props.putData('/', {}, '');

    expect(dispatch).toHaveBeenCalledWith(expect.any(Function));
  });

  it('should dispatch setAuthentication when setAuthentication is called', () => {
    const dispatch = jest.fn();
    const props = mapDispatchToProps(dispatch);
    props.setAuthentication({}, 'test');

    expect(dispatch).toHaveBeenCalledWith({ auth: {}, type: 'SET_AUTH' });
  });

  it('should dispatch setLocation when setLocation is called', () => {
    const dispatch = jest.fn();
    const props = mapDispatchToProps(dispatch);
    props.setLocation('home');

    expect(dispatch).toHaveBeenCalledWith({ location: 'home', type: 'SET_LOCATION' });
  });
});

describe('mapStateToProps', () => {
  it('should return the correct location', () => {
    const state = {
      location: 'signup'
    };

    expect(mapStateToProps(state)).toEqual({
      location: 'signup'
    });
  });
});

describe('validatorTest', () => {
  const dummyValues = {
    email: 'test@gmail.com',
    password: 'password test',
    cpassword: 'password test'
  };

  it('should validate email is required', () => {
    expect(validate({ ...dummyValues, email: null })).toEqual({
      email: 'Required'
    });
  });

  it('should validate email is a real email', () => {
    expect(validate({ ...dummyValues, email: 'not an email' })).toEqual({
      email: 'Invalid email address'
    });
  });

  it('should validate password is required', () => {
    expect(validate({ ...dummyValues, password: null, cpassword: null })).toEqual({
      password: 'Required'
    });
  });

  it('should validate password is less than 16 characters', () => {
    expect(
      validate({
        ...dummyValues,
        password: 'the quick brown fox jumps over the lazy dog',
        cpassword: 'the quick brown fox jumps over the lazy dog'
      })
    ).toEqual({
      password: 'Must be 6-16 characters'
    });
  });

  it('should validate password is more than 6 characters', () => {
    expect(
      validate({
        ...dummyValues,
        password: 'hi',
        cpassword: 'hi'
      })
    ).toEqual({
      password: 'Must be 6-16 characters'
    });
  });

  it('should validate password is the same as cpassword', () => {
    expect(
      validate({
        ...dummyValues,
        password: 'hi buddy',
        cpassword: 'not hi buddy'
      })
    ).toEqual({
      cpassword: 'Passwords must match'
    });
  });
});
