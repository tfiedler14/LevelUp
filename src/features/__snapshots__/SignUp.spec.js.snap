// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn should render sign up correctly 1`] = `
<ImageBackground
  source={1}
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "paddingTop": 100,
      }
    }
  >
    <withTheme(Card)>
      <View>
        <FormHeader
          title="Sign Up"
        />
        <Field
          component={[Function]}
          id="email"
          name="email"
          props={
            Object {
              "autoCompleteType": "email",
              "textContentType": "emailAddress",
              "title": "Email",
            }
          }
        />
        <Field
          component={[Function]}
          id="password"
          name="password"
          props={
            Object {
              "autoCompleteType": "password",
              "password": true,
              "textContentType": "password",
              "title": "Password",
            }
          }
        />
        <Field
          component={[Function]}
          id="cpassword"
          name="cpassword"
          props={
            Object {
              "autoCompleteType": "password",
              "password": true,
              "textContentType": "password",
              "title": "Confirm Password",
            }
          }
        />
        <View>
          <withTheme(Button)
            color="#de3c57"
            mode="contained"
            uppercase={false}
          >
            Sign Up
          </withTheme(Button)>
          <withTheme(Button)
            color="white"
            mode="text"
            onPress={[Function]}
            uppercase={false}
          >
            Sign In
          </withTheme(Button)>
        </View>
      </View>
    </withTheme(Card)>
  </View>
</ImageBackground>
`;
